### PR TITLE
Release process fix & major/minor/rev macro addition

### PR DESCRIPTION
**Pull request by d-a-v**
- Release process update is based on #8034.
- `makecorever.py` is now used as part of the release process script.
- `MAJOR`/`MINOR`/`REVISION` are added in `core_version.h` based on `platform.txt`'s `version` when building with Arduino-CLI, or git tag when releasing. PlatformIO will have to be updated too (this is not addressed here).

This is WIP but is going to be merged anyway in order to try the release script executed remotely by CI.
A tag will be added (3.0.1) and removed right after that (this gymnastic happens quite often while releasing these days).
This will allow to check generated logs and archive to address #8049

### CODE REVIEW

File: tools/makecorever.py [hunk 1]
```diff
@@ -20,23 +20,40 @@
...
 
-def generate(path, platform_path, git_ver="ffffffff", git_desc="unspecified"):
+def generate(path, platform_path, git_ver="ffffffff", platform_version="unspecified"):
     def git(*args):
         cmd = ["git", "-C", platform_path]
         cmd.extend(args)
         proc = subprocess.Popen(cmd, stdout=subprocess.PIPE, universal_newlines=True, stderr=subprocess.DEVNULL)
         return proc.stdout.readlines()[0].strip()
 
+    git_desc = platform_version;
```
Comment (mcspr) [inline tools/makecorever.py:33]: Following the above, should this have additional param to skip git-describe?
Meaning, this still could still work with a shallow copy or with some broken tagging (like github actions checkout does)

File: tools/makecorever.py [hunk 2]
```diff
@@ -20,23 +20,40 @@
...
+    text += "#define ARDUINO_ESP8266_MINOR     {}\n".format(version[1])
+    # revision can be ".n" or ".n-dev"
+    revision = re.split("-", version[2])
+    text += "#define ARDUINO_ESP8266_REVISION  {}\n".format(revision[0])
+    text += "\n"
+    if len(revision) > 1:
+        text += "#define ARDUINO_ESP8266_DEV       1 // developpment version\n"
+    else:
+        text += "#define ARDUINO_ESP8266_RELEASE   \"{}\"\n".format(git_desc)
+        text += "#define ARDUINO_ESP8266_RELEASE_{}\n".format(re.sub("[-\.]", "_", git_desc))
```
Comment (mcspr) [inline tools/makecorever.py:56]: also, re. `re`
python has string methods like `"1.2.3".split(".")` and `"1.2.3".replace(".", "_")`
also note of the `"1.2.3-dev".partition("-")`

File: package/build_boards_manager_package.sh [hunk 3]
```diff
@@ -109,12 +109,13 @@ $SED -E "s/name=([a-zA-Z0-9\ -]+).*/name=\1(${ver})/g"\
...
-echo "#define ARDUINO_ESP8266_GIT_DESC `git describe --tags 2>/dev/null`" >>${outdir}/cores/esp8266/core_version.h
-echo "#define ARDUINO_ESP8266_RELEASE_${ver_define}" >>${outdir}/cores/esp8266/core_version.h
-echo "#define ARDUINO_ESP8266_RELEASE \"${ver_define}\"" >>${outdir}/cores/esp8266/core_version.h
+#ver_define=`echo "${plain_ver}" | tr "[:lower:]." "[:upper:]_"`
+#echo "ver_define: ${ver_define}  (plain_ver: ${plain_ver})"
+#echo "#define ARDUINO_ESP8266_GIT_VER 0x`git rev-parse --short=8 HEAD 2>/dev/null`" >${outdir}/cores/esp8266/core_version.h
+#echo "#define ARDUINO_ESP8266_GIT_DESC `git describe --tags 2>/dev/null`" >>${outdir}/cores/esp8266/core_version.h
+#echo "#define ARDUINO_ESP8266_RELEASE_${ver_define}" >>${outdir}/cores/esp8266/core_version.h
+#echo "#define ARDUINO_ESP8266_RELEASE \"${ver_define}\"" >>${outdir}/cores/esp8266/core_version.h
+python3 ${srcdir}/tools/makecorever.py -b ${outdir} -i cores/esp8266 -p ${srcdir} -v ${plain_ver}
```
Comment (mcspr) [inline package/build_boards_manager_package.sh:118]: Should the script have an explicit option to toggle the release mode, so we don't have to auto-detect?

File: tools/makecorever.py (hunk 1 above)
Comment (d-a-v) [reply tools/makecorever.py]: Tagging works OK including when used by github-action. Released are based on tagging.
I updated the script per your 3 reviews and added comments.

Comment (mcspr) [reply tools/makecorever.py]: broken == checkout doing `git fetch origin +$commit-sha$:refs/tags/$tag-name$`, which is worked around with fetch-depth:0... which could be avoided

my only problem with that is that version string depends on too much stuff, where we already know it from the tag name anyway; and commit sha comes from rev-parse regardless, shallow copy or not.

**Pull request merged**